rclcpp/include/rclcpp/node_interfaces/node_base.hpp

```diff
@@ -0,0 +1,118 @@
+// Copyright 2016 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__NODE_INTERFACES__NODE_BASE_HPP_
+#define RCLCPP__NODE_INTERFACES__NODE_BASE_HPP_
+
+#include "rclcpp/context.hpp"
+#include "rclcpp/macros.hpp"
+#include "rclcpp/node_interfaces/node_base_interface.hpp"
+#include "rclcpp/visibility_control.hpp"
+
+namespace rclcpp
+{
+namespace node_interfaces
+{
+
+/// Implementation of the NodeBaseInterface, which is part of the Node API.
+class NodeBase : public NodeBaseInterface
+{
+public:
+  RCLCPP_SMART_PTR_ALIASES_ONLY(NodeBaseInterface)
+
+  RCLCPP_PUBLIC
+  NodeBase(const std::string & node_name, rclcpp::context::Context::SharedPtr context);
+
+  RCLCPP_PUBLIC
+  virtual
+  ~NodeBase();
+
+  RCLCPP_PUBLIC
+  virtual
+  const std::string &
+  get_name() const;
+
+  RCLCPP_PUBLIC
+  virtual
+  rclcpp::context::Context::SharedPtr
+  get_context() const;
```

Comment:
I don't know. I didn't make this `const` originally (this method was just moved not modified or added) because the shared ptr that is returned is a copy. From what I understand it is not required and not appropriate to return a const of a copy, e.g. `bool get_x() const` and not `const bool get_x() const`. Instead you return a const when you return a reference, e.g. `const MyClass & get_my_class() const` rather than `MyClass & get_my_class() const`.

So if you consider the shared pointer to be a copy, then it doesn't need to be const, i.e. if someone changes the shared ptr that was returned, or calls reset of on it, it would not change the class. However, since it is a pointer, you could modify the thing to which it points and so maybe it should be a `::ConstSharedPtr` as we call it (aka `std::shared_ptr<const Context>`). Getting rid of the shared pointer and just using a raw pointer, it is currently `Context * get_context() const` and maybe it should be `Context const * get_context() const` (reads "the const method get_context returns a pointer to a const Context").

I'm not sure there is a needs to be a const and non const version. I think a non const version is fine, and if the method could be const then that would be better, but not necessary.

So the question is whether or not we should leave it as is, or if we should have it return a `SharedPtr` but not be a const method, or if we should have it return a `ConstSharedPtr`and be a const method.